libbpf-tools/javagc.c

```diff
@@ -0,0 +1,224 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/*
+ * Based on ugc from BCC by Sasha Goldshtein
+ * Author: Chen Tao
+ * Create: Wed Jun 29 16:00:19 2022
+ */
+#include <stdio.h>
+#include <argp.h>
+#include <signal.h>
+#include <unistd.h>
+#include <time.h>
+#include <sys/resource.h>
+#include <bpf/libbpf.h>
+#include <errno.h>
+#include "javagc.skel.h"
+#include "javagc.h"
+
+#define BINARY_PATH_SIZE (256)
+#define PERF_BUFFER_PAGES (32)
+#define PERF_POLL_TIMEOUT_MS (200)
+
+static struct env {
+	pid_t pid;
+	int time;
+	bool exiting;
+	bool verbose;
+} env = {
+	.pid = -1,
+	.time = 1000,
+	.exiting = false,
+	.verbose = false,
+};
+
+const char *argp_program_version = "javagc 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+
+const char argp_program_doc[] =
+"Monitor javagc time cost.\n"
+"\n"
+"USAGE: javagc [--help] [-p PID] [-t GC time]\n"
+"\n"
+"EXAMPLES:\n"
+"javagc -p 185         # trace PID 185 only\n"
+"javagc -p 185 -t 100  # trace PID 185 java gc time beyond 100us\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "time", 't', "TIME", 0, "Java gc time" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int err = 0;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno) {
+			err = errno;
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 't':
+		errno = 0;
+		env.time = strtol(arg, NULL, 10);
+		if (errno) {
+			err = errno;
+			fprintf(stderr, "invalid time: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return err;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && ! env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct data_t *e = (struct data_t *)data;
+	struct tm *tm = NULL;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	printf("time:%s, cpu:%d, pid:%d, gc time(us):%llu\n", ts, e->cpu,
```

Comment:
Add a space after `:` here and below for the error messages.

Reply:
Fix it